### PR TITLE
Allow build with nix-tools.

### PR DESCRIPTION
**Pull request by jbgi**
Needs to be done in this PR:
- [x] Hydra jobset for Linux.
- [x] Hydra jobset for MacOS.
- [x] Builds of cardano-http-bridge, jörmungandr, cardano-sl-node on linux.
- [x] Nix-shell development environment with all dependencies installed.
- [x] Buildkite job to check hydra evaluation. @cleverca22 
- [x] Buildkite job to regenerate nix and provide diff. input-output-hk/iohk-nix#88

Can do it later:
- [ ] Cross builds for windows - some dependencies are not building yet. @angerman 
- [ ] Benchmarks building @angerman @rvl
- [ ] Builds of cardano-http-bridge, jörmungandr on MacOS. @cleverca22 @disassembler 
- [ ] Integration tests passing.

[Hydra jobset](https://hydra.iohk.io/jobset/Cardano/cardano-wallet-pr-103)

### CODE REVIEW

**Comment by KtorZ** [comment]
May I ask, what's the rational here :no_mouth: ?

**Comment by jbgi** [comment]
I would say cross-compilation support and better caching are the main reasons.

File: stack.yaml [hunk 1]
```diff
@@ -1,4 +1,5 @@
 resolver: lts-13.8
+compiler: ghc-8.6.3
```
Comment (rvl) [inline stack.yaml]: Why is this needed, if LTS 13.8 already implies GHC 8.6.3?

File: nix/stack-shell.nix [hunk 2]
```diff
@@ -3,7 +3,7 @@
 { system ? builtins.currentSystem
 , config ? {}
 , walletPackages ? import ./.. { inherit config system; }
-, pkgs ? walletPackages.pkgs
+, pkgs ? walletPackages._lib.pkgs
```
Comment (rvl) [inline nix/stack-shell.nix]: Can `iohk-nix` be improved here? Getting the pinned pkgs seems like a fairly standard thing to do, so there should be a "standard" way to do it.

File: release.nix [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  #   nix-tools.libs.x86_64-pc-mingw32-iohk-monitoring.x86_64-linux -- will build the iohk-monitoring library on linux for windows.
+  #   nix-tools.tests.iohk-monitoring.tests.x86_64-linux -- will build and run the tests from the
+  #                                                          iohk-monitoring package on linux.
+
+  # The required jobs that must pass for ci not to fail:
+  required-name = "cardano-wallet-required-checks";
+  required-targets = jobs: [
+    # targets are specified using above nomenclature:
+    jobs.nix-tools.tests.cardano-wallet.unit.x86_64-linux
+    jobs.nix-tools.tests.cardano-wallet.integration.x86_64-linux
```
Comment (rvl) [inline release.nix]: As they are, the integration tests probably won't work in a sandboxed build.

File: release.nix [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    jobs.nix-tools.exes.cardano-wallet.x86_64-linux
```
Comment (rvl) [inline release.nix]: The wallet needs `cardano-http-bridge` in its `PATH` - can something be done about it?

File: default.nix [hunk 5]
```diff
@@ -1,20 +1,50 @@
...
+# This file needs to export a function that takes
+# the arguments it is passed and forwards them to
+# the default-nix template from iohk-nix. This is
+# important so that the release.nix file can properly
+# parameterize this file when targetting different
+# hosts.
+{ ... }@args:
+# We will instantiate the default-nix template with the
+# nix/pkgs.nix file...
+commonLib.nix-tools.default-nix ./nix/pkgs.nix args
```
Comment (rvl) [inline default.nix]: This `commonLib.nix-tools.default-nix` function is very obscure. I would prefer it if `default.nix` did not use that function, even if it means that some code has been repeated.

File: nix/stack-shell.nix [hunk 6]
```diff
@@ -3,7 +3,7 @@
 { system ? builtins.currentSystem
```
Comment (rvl) [inline nix/stack-shell.nix]: When I try `stack --nix build`, it starts to build GHC 8.4.3 (`/nix/store/7sih6g9zyzbrjc0qqgq5692m3pwgazkz-ghc-8.4.3.drv`). Why is that?

**Review by rvl — DISMISSED** [review]
Thanks for working on this @jbgi.

Does the windows build work?

**Comment by KtorZ** [comment]
I wonder: is there a way to get cross-compilation _without_ nix? We aren't going to use nix on our CI (or get the whole team trained for that before, but I don't see this happening in the near future :/) but we might want to get cross-compilation to windows?

**Comment by rvl** [comment]
@KtorZ I am more than happy to provide training in Nix. It won't take long to learn the basics.
Cross-compilation (and other things) would be a lot harder without Nix.

**Comment by KtorZ** [comment]
Closing until we decide to allocate resources on this particular feature / tooling.

File: default.nix [hunk 7]
```diff
@@ -1,20 +1,56 @@
...
-# Pin nixpkgs to a revision on the nixos-19.03 channel
-, nixpkgs ? iohkLib.fetchNixpkgs ./nix/nixpkgs-src.json
-, pkgs ? import nixpkgs { inherit system crossSystem config; }
+, iohkLib ? import ./nix/iohk-common.nix { inherit system crossSystem config; }
+# Use nixpkgs pin from iohkLib
+, pkgs ? iohkLib.pkgs
 
 # Keep this argument even if unused.
 # It will prevent Hydra from caching the evaluation.
 , gitrev ? iohkLib.commitIdFromGitRepo ./.
```
Comment (cleverca22) [inline default.nix]: this will only trigger a new eval if its actually an input to a job

it needs to be routed to the forceNewEval job that iohk-nix is generating

**Review by cleverca22 — DISMISSED** [review]

File: default.nix (hunk 7 above)
Comment (rvl) [reply default.nix]: Good point. I have removed it from here, and made it local to the `required` aggregate job.

**Review by rvl — APPROVED** [review]

**Review by KtorZ — APPROVED** [review]
